### DISCUSSION TITLE
Best place to make API calls

### DISCUSSION THREAD

**Discussion by capoia** [GrapesJS/grapesjs#5612] [Q&A]
Hello, I'm trying to build a component that connects to my CRM, and before building the component I need to make some api calls, I'm using ofetch for this.
The problem is, I need to use async and await, however if I put the export default which is inside index.js (in grapesjs-cli) it randomly gives the plugin loading error, that is, I update the page 5 times, 4 work and one of the component definition error: "Plugin cube-form not found {context: 'plugins', plugin: 'cubo-form', level: 'warning'}".

I realized that removing async from the default export solves the problem, but the question remains, where do I make the calls to the api? I even tried to do it outside the export default, but I need the opts that are inside the export default (to get the token that the user enters to make the api calls).

![CleanShot 2024-01-11 at 15 49 45](https://github.com/GrapesJS/grapesjs/assets/21159928/93753c73-6c99-4d82-8fb5-d0fac086a127)
![CleanShot 2024-01-11 at 15 50 08](https://github.com/GrapesJS/grapesjs/assets/21159928/01500a06-331d-4c24-bc44-66fc509446bf)

**Comment by artf** [comment]
Can you please elaborate better your use case, what is the purpose of the async call inside the plugin?

**Reply by capoia** [reply]
My component is a form, which when sent sends the lead to my CRM, within my CRM I have funnels and stages.
Within the traits of my component there are selects for funnels and stages, I need to pull these funnels and stages from my crm, to list them in the traits, for the user to select and when submitting the form they go to the funnel and stage they would like.

**Reply by capoia** [reply]
and in this case I cannot make the call outside of export because I need the user's token (these routes for recovering funnels and stages are protected by the token).
The token is an attribute of the plugin and I get it when the person enters the pagebuilder and send it to the component to use in these calls.

**Reply by artf** [reply]
You could place that call in [Component View init hook](https://grapesjs.com/docs/modules/Components.html#lifecycle-hooks) and after getting the data, [update the trait](https://grapesjs.com/docs/api/component.html#gettrait) options.
